### ISSUE TITLE
`docker-compose run` does not detect EOF and hangs

### ISSUE THREAD

**Issue by thom-nic** [docker/compose#3829]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        Thu Jul 28 21:15:28 2016
 OS/Arch:      darwin/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:        
 OS/Arch:      linux/amd64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Using physical box (CentOS 7 host) via docker-machine or docker for mac, results are the same.

**Steps to reproduce the issue:**
1. Using docker-compose specify a named service (the base image does not seem to matter)
2. run either of the following:

```
docker-compose run [service] cat < somefile.txt
# or :
echo 'hello there' |docker-compose run [service] cat
```

**Describe the results you received:**
Client hangs and does not exit after contents of redirected input are consumed:

```
± docker-compose run web cat < docker/postgres/grafana.sql
CREATE ROLE grafana LOGIN PASSWORD 'grafana';
CREATE DATABASE grafana;
GRANT ALL PRIVILEGES ON DATABASE grafana TO grafana;
^C^C^CERROR: Aborting.

± echo 'hello there' |docker-compose run web cat
hello there
^C^CERROR: Aborting.
```

**Describe the results you expected:**
Client should exit after redirected stdin hits EOF.  The same behavior with `docker run` exits cleanly as expected:

```
± echo 'hello there' |docker run --rm -i alpine:latest cat
hello there

± docker run --rm -i alpine:latest cat < docker/postgres/grafana.sql
CREATE ROLE grafana LOGIN PASSWORD 'grafana';
CREATE DATABASE grafana;
GRANT ALL PRIVILEGES ON DATABASE grafana TO grafana;

± 
```

**Describe any additional info:**

I've noticed similar behavior when doing `docker-compose run [servicename] /bin/sh` - when I attempt to exit the shell (with `exit` or `^D`) the client doesn't exit until I hit `^C`:

```
± docker-compose run web /bin/sh
~ $ 
~ $ exit
^C^C^CERROR: Aborting.
```

**Comment by zombie-guru** [comment]
Having the same issue with docker-machine. docker-compose run exits cleanly on EOF when running docker on linux though.

**Comment by thom-nic** [comment]
I don't doubt this issue could be specific to the mac client, I haven't tried from a linux client for comparison.

**Comment by gyp** [comment]
It happens for me, too. My guess that it is a packaging issue because I can reproduce the problem with a pip-installed version but it works fine with the very same build that I download as a binary, all at the same time on the very same Ubuntu 16.04:

```
# curl -s -L "https://github.com/docker/compose/releases/download/1.9.0/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose-binary
# pip install -q docker-compose

# docker-compose version
docker-compose version 1.9.0, build 2585387
docker-py version: 1.10.6
CPython version: 3.5.2
OpenSSL version: OpenSSL 1.0.2g  1 Mar 2016

# docker-compose-binary version
docker-compose version 1.9.0, build 2585387
docker-py version: 1.10.6
CPython version: 2.7.9
OpenSSL version: OpenSSL 1.0.1t  3 May 2016

# docker version
Client:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:19:35 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.2
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   bb80604
 Built:        Tue Oct 11 18:19:35 2016
 OS/Arch:      linux/amd64
```

The simplest way to reproduce the bug for me was the echo | docker-compose | cat method used above:

```
# cat docker-compose.yml 
version: '2'

services:
  ubuntu:
    image: ubuntu
    command: /bin/bash

# echo "hello world" | docker-compose run --rm ubuntu cat
hello world
^C^C#

# echo "hello world" | docker-compose-binary run --rm ubuntu cat
hello world
# 
```
Note that the second call using the prepackaged docker-compose works just fine. Calling docker directly works perfectly:

```
# echo "hello world" | docker run -i ubuntu cat
hello world
#
```

In more complex situations when the consumer within the container was a python app that parsed its stdin as a CSV I saw it get stuck earlier and not at the EOF. Weirdly enough, at some random locations, if I cut the CSV input down to, say 10k lines then at 9500 lines, when it was 1000 lines then somewhere around 700 and when it was only 500 lines then at 450 -- but always at the same point, repeatedly. Unfortunately I couldn't find an easy self-contained way to reproduce the latter issue.

**Comment by shin-** [comment]
Fixed by #5580
